Automattic/woocommerce-payments
# Apple Pay domain verification

Fixes #1071

Child of #1175

Depends on [Automattic/woocommerce-payments-server#535](https://github.com/Automattic/woocommerce-payments-server/pull/535)

#### Changes proposed in this Pull Request
* Adapt Apple pay registration module from Stripe Gateway extension.
  * Add rewrite rule for `/.well-known/apple-developer-merchantid-domain-association`.
  * Add `apple-developer-merchantid-domain-association` file copy as a fallback for verification.
* Add server API call to request verification in Stripe.
* Add `woocommerce_woocommerce_version` option.
* Add update routine to update domain association file and trigger domain registration.
* Verify domain automatically upon domain name change.

#### Testing instructions


**Before you begin**

* This PR [Automattic/woocommerce-payments-server#535](https://github.com/Automattic/woocommerce-payments-server/pull/535) needs to be merged into trunk already, or you need to checkout into that branch first.
* The domain registration requires a live connected account and a live secret key. You can use WooCommerce Payments live public key.
* For security reasons I'm not going to post any Stripe credentials here. Ping me in Slack (@ricardometring) so I can give you the live restricted key that only works for Apple Pay Domains registration for the purpose of this test - and you may also connect to my own personal live connected account.
* I recommend using Ngrok instead of Jurassic tube for this test, because you can test multiple subdomains more easily with Ngrok's free tier by simply reconnecting it.

**How to test**
1. Start the client and server docker environments.
2. Start Ngrok or Jurassic Tube to get a public URL.
3. Turn off Dev mode and API request redirection in WCPay Dev Utils.
4. Start the Jetpack onboarding process and make sure your account is connected. - When you see the Stripe onboarding screen, **don't** continue and go to wp-admin instead.
5. Go to the WCPay Dev Utils plugin, grab your WP.com blog ID and connect with the account I sent you over Slack using `/local/bin/link-account.sh <blog-id> <stripe-account-id> live` - or your own live account if you have one.
6. Enable API request redirection to your server's address (Keep dev mode disabled).
7. Update your secrets.php with the live account info I sent you over Slack, in case you haven't already.

8. There are 3 things that can trigger domain registration with Apple. You can try either one or all of them:
  * Delete the option `woocommerce_woocommerce_payments_version` and visit any wp-admin page to simulate a plugin update.
  * Disable the Payment Request button in `Payments > Settings`, save, enable it, then save again.
  * Change your domain and visit any wp-admin page.

  You don't necessarily need any of the steps above if you manage to connect the live account before accessing wp-admin. This is how it's supposed to work for our customers.

9. Open your HTTPS Ngrok or Jurassic Tube URL in Safari and navigate to a product page. If you have Apple Wallet enabled with a verified card, you should be able to see the Apple Pay Button.

10. To test the payment, you need to disable API request redirection, because that live secret key is only enabled for registering a domain, not capturing payments. Make sure you have test mode enabled, otherwise you'll be charged.

#### Screenshots

  <img width="300" alt="Apple Pay Button" src="https://user-images.githubusercontent.com/5509901/105923342-fde52080-601a-11eb-93f5-3df57fe49beb.png"> <img width="200" alt="Apple Pay Button" src="https://user-images.githubusercontent.com/5509901/105925286-992bc500-601e-11eb-95a0-f07c9fdb0843.PNG">

Buy - Dark / Branded - Text and logo

<img width="422" alt="Screen Shot 2021-01-30 at 12 47 15" src="https://user-images.githubusercontent.com/5509901/106361506-0c725700-62fd-11eb-9c62-b7319808d3b8.png">

Donate - Dark

<img width="422" alt="Screen Shot 2021-01-30 at 12 47 40" src="https://user-images.githubusercontent.com/5509901/106361511-13996500-62fd-11eb-96f0-584863b028ef.png">

Buy - Light

<img width="415" alt="Screen Shot 2021-01-30 at 12 46 55" src="https://user-images.githubusercontent.com/5509901/106361536-404d7c80-62fd-11eb-9069-e47bafe18731.png">

Buy - Light-Outline

<img width="424" alt="Screen Shot 2021-01-30 at 12 46 23" src="https://user-images.githubusercontent.com/5509901/106361537-43e10380-62fd-11eb-813e-7a71604ce374.png">

Branded - Logo only

<img width="422" alt="Screen Shot 2021-01-30 at 12 48 12" src="https://user-images.githubusercontent.com/5509901/106361518-26ac3500-62fd-11eb-948e-805ea584ba73.png">


-------------------

- [x] Added changelog entry - does not apply - Added in parent PR
